File: app/services/welcome.py
from __future__ import annotations

from aiogram.types import InputRichMessage, User

from app.i18n import t

SHOWCASE_URL = "https://t.me/TelegramTips/573"
UPDATES_URL = "https://t.me/RichCustomizet"
SUPPORT_URL = "https://t.me/+D4cEzE0V7IIwYTcx"
BOT_USERNAME = "@RichCustomizebot"


def _url_button(text: str, url: str) -> dict[str, object]:
    return {
        "type": "button",
        "button": {
            "text": text,
            "url": url,
        },
    }


def build_welcome_rich_message(user: User) -> InputRichMessage:
    """Build the /start message using Telegram RichText buttons and a real user mention."""
    mention_text = user.full_name.strip() or str(user.id)
    text: list[object] = [
        f"{t('welcome.greeting')} ",
        {
            "type": "text_mention",
            "text": mention_text,
            "user": user,
        },
        f"!\n{BOT_USERNAME}\n",
        t("welcome.description_before_view"),
        _url_button(t("welcome.view_button"), SHOWCASE_URL),
        t("welcome.description_after_view"),
        "\n\n",
        t("welcome.add_prompt"),
        "\n\n",
        t("welcome.help_title"),
        "\n",
        f"{t('welcome.help_check')} ",
        _url_button(t("welcome.updates_button"), UPDATES_URL),
        " && ",
        _url_button(t("welcome.support_button"), SUPPORT_URL),
    ]
    return InputRichMessage(blocks=[{"type": "paragraph", "text": text}])


__all__ = [
    "BOT_USERNAME",
    "SHOWCASE_URL",
    "SUPPORT_URL",
    "UPDATES_URL",
    "build_welcome_rich_message",
]

File: lib/welcome.js
import { isRtlLocale, resolveLanguage, t } from 'lib/i18n';

export const SHOWCASE_URL = 'https://t.me/durov/531';
export const UPDATES_URL = 'https://t.me/RichCustomize';
export const SUPPORT_URL = 'https://t.me/+D4cEzE0V7IIwYTcx';
export const ADD_GROUP_URL = 'https://t.me/RichCustomizebot?startgroup=true';
export const BOT_USERNAME = '@RichCustomizebot';

function urlButton(text, url) {
  return {
    type: 'button',
    button: { text, url },
  };
}

function userDisplayName(user) {
  if (!user) return '';
  const name = [user.first_name, user.last_name].filter(Boolean).join(' ').trim();
  return name || String(user.id || '');
}

export function buildWelcomeRichMessage(user, languageCode) {
  const locale = resolveLanguage(languageCode);
  const mentionText = userDisplayName(user);
  const greeting = [t(locale, 'welcome.greeting') + ' '];

  if (user?.id && mentionText) {
    greeting.push({
      type: 'text_mention',
      text: mentionText,
      user,
    });
  } else if (mentionText) {
    greeting.push(mentionText);
  }
  greeting.push('!');

  const richMessage = {
    blocks: [
      { type: 'heading', text: greeting, size: 3 },
      { type: 'footer', text: '- ' + BOT_USERNAME },
      {
        type: 'paragraph',
        text: [
          t(locale, 'welcome.product_description'),
          ' ',
          urlButton(t(locale, 'welcome.view_button'), SHOWCASE_URL),
        ],
      },
      { type: 'divider' },
      {
        type: 'footer',
        text: [
          t(locale, 'welcome.help_title'),
          '\n',
          t(locale, 'welcome.help_check') + ' ',
          urlButton(t(locale, 'welcome.updates_button'), UPDATES_URL),
          ' ' + t(locale, 'welcome.and') + ' ',
          urlButton(t(locale, 'welcome.support_button'), SUPPORT_URL),
        ],
      },
    ],
  };

  if (isRtlLocale(locale)) richMessage.is_rtl = true;
  return richMessage;
}

export function buildWelcomeKeyboard(languageCode) {
  const locale = resolveLanguage(languageCode);
  return {
    inline_keyboard: [
      [{ text: t(locale, 'welcome.add_group_button'), url: ADD_GROUP_URL }],
      [
        {
          text: t(locale, 'welcome.showcase_button'),
          callback_data: 'r:showcase',
        },
        {
          text: t(locale, 'editor.new_button'),
          callback_data: 'r:starteditor',
          style: 'primary',
        },
      ],
    ],
  };
}

export function buildWelcomeFallbackText(user, languageCode) {
  const locale = resolveLanguage(languageCode);
  const name = userDisplayName(user);
  return [
    t(locale, 'welcome.greeting') + (name ? ' ' + name : '') + '!',
    '',
    t(locale, 'welcome.product_description'),
    '',
    t(locale, 'welcome.help_title'),
    t(locale, 'welcome.help_check') + ' ' + UPDATES_URL + ' ' + t(locale, 'welcome.and') + ' ' + SUPPORT_URL,
  ].join('\n');
}
